akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -313,6 +316,88 @@ private[stream] class VirtualPublisher[T]() extends Publisher[T] {
 /**
  * INTERNAL API
  */
+private[stream] case class MaterializedValueSource[M](
+  shape: SourceShape[M] = SourceShape[M](new Outlet[M]("Materialized.out")),
+  attributes: OperationAttributes = OperationAttributes.name("Materialized")) extends StreamLayout.Module {
+
+  override def subModules: Set[Module] = Set.empty
+  override def withAttributes(attr: OperationAttributes): Module = this.copy(shape = amendShape(attr), attributes = attr)
+  override def carbonCopy: Module = this.copy(shape = SourceShape(new Outlet[M]("Materialized.out")))
+
+  override def replaceShape(s: Shape): Module =
+    if (s == shape) this
+    else throw new UnsupportedOperationException("cannot replace the shape of MaterializedValueSource")
+
+  def amendShape(attr: OperationAttributes): SourceShape[M] = {
+    attr.nameOption match {
+      case None ⇒ shape
+      case s: Some[String] if s == attributes.nameOption ⇒ shape
+      case Some(name) ⇒ shape.copy(outlet = new Outlet(name + ".out"))
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] object MaterializedValuePublisher {
+  val NotRequested = 0
+  val Requested = 1
+  val Completed = 2
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] class MaterializedValuePublisher extends Publisher[Any] {
+  import MaterializedValuePublisher._
+
+  private val value = new AtomicReference[AnyRef](null)
+  private val registeredSubscriber = new AtomicReference[Subscriber[_ >: Any]](null)
+  private val requestState = new AtomicInteger(NotRequested)
+
+  def setValue(m: Any): Unit =
+    if (value.compareAndSet(null, m.asInstanceOf[AnyRef]) && requestState.get() == Requested)
+      pushAndClose(m)
+
+  private def pushAndClose(m: Any): Unit = {
+    val sub = registeredSubscriber.get()
+    ReactiveStreamsCompliance.tryOnNext(sub, m)
+    ReactiveStreamsCompliance.tryOnComplete(sub)
+    requestState.set(Completed)
```

Comment:
this needs to be a compareAndSet and should guard the onNext/onComplete calls to avoid duplicate signaling. Not that it can happen right now, since materialization is synchronous (the whole atomic stuff is strictly unnecessary), but hey.